src/publisher/ajson_ingestor.py

```diff
@@ -31,51 +31,82 @@
+    # this *could* be scraped from the provided data, but we have no time to
+    # normalize journal names so we sometimes get duplicate journals in the db.
+    # safer to disable until needed.
+    journal = logic.journal()
 
-        try:
-            article_struct = render.render_item(ARTICLE, data['article'])
-            article, created, updated = \
-                create_or_update(models.Article, article_struct, ['manuscript_id', 'journal'], create, update, journal=journal)
+    try:
+        article_struct = render.render_item(ARTICLE, data['article'])
+        article, created, updated = \
+            create_or_update(models.Article, article_struct, ['manuscript_id', 'journal'], create, update, journal=journal)
 
-            assert isinstance(article, models.Article)
-            log_context['article'] = article
+        assert isinstance(article, models.Article)
+        log_context['article'] = article
 
-            previous_article_versions = None
-            if updated:
-                previous_article_versions = list(article.articleversion_set.all().order_by('version')) # earliest -> latest
+        previous_article_versions = []
+        if updated:
+            previous_article_versions = list(article.articleversion_set.all().order_by('version')) # earliest -> latest
 
-            av_struct = render.render_item(ARTICLE_VERSION, data['article'])
-            # this is an INGEST event and *not* a PUBLISH event. we don't touch the date published.
-            del av_struct['datetime_published']
+        av_struct = render.render_item(ARTICLE_VERSION, data['article'])
+        # this is an INGEST event and *not* a PUBLISH event. we don't touch the date published.
+        del av_struct['datetime_published']
 
-            av, created, updated = \
-                create_or_update(models.ArticleVersion, av_struct, ['article', 'version'],
-                                 create, update, commit=False, article=article)
-        except KeyError as err:
-            raise ValueError("failed to scrape article data, couldn't find key %s" % err)
+        av, created, updated = \
+            create_or_update(models.ArticleVersion, av_struct, ['article', 'version'],
+                             create, update, commit=False, article=article)
 
         assert isinstance(av, models.ArticleVersion)
         log_context['article-version'] = av
 
+        data['article']['forced?'] = force
+        ae_structs = [render.render_item(desc, data['article']) for desc in INGEST_EVENTS]
+        # ignore any events we don't have explicit datetimes for
+        [events.add(article, **struct) for struct in ae_structs if 'datetime_event' in struct]
+
+        # return journal, article, av, ael
+        return av, created, updated, previous_article_versions
+
+    except KeyError as err:
+        raise ValueError("failed to scrape article data, couldn't find key %s" % err)
+
+
+#
+#
+#
+
+def _ingest(data, force=False) -> models.ArticleVersion:
```

Comment:
this is interesting. a few weeks ago I came across typing in python 3 that had just never crossed my radar before. I'm looking for a static analysis tool that I can run as part of the linting and possibly replace pylint. `mypy` is a strong contender: http://mypy-lang.org/